Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,501 @@
+        simp only [deriv_id'', one_div]
+        · have {q : ℝ} (p : ℝ) : DifferentiableAt ℝ (fun p => q - p) p := by fun_prop
+          have d_oneminus (x : ℝ) : deriv (fun (y : ℝ) ↦ 1 - y) x = -1 := by
+            rw [deriv_const_sub 1, deriv_id'']
+          field_simp [sub_ne_zero_of_ne xne1.symm, this, d_oneminus]
+          ring
+      · apply DifferentiableAt.add
+        simp only [ne_eq, differentiableAt_const]
+        exact DifferentiableAt.log (by fun_prop) (sub_ne_zero.mpr xne1.symm)
+    filter_upwards [eventually_ne_nhds xne0, eventually_ne_nhds xne1]
+      with y xne0 h2 using deriv_qaryEntropy xne0 h2
+  -- Pathological case where we use junk value (because function not differentiable)
+  · have : x = 0 ∨ x = 1 := Decidable.or_iff_not_and_not.mpr is_x_where_nondiff
+    rw [deriv_zero_of_not_differentiableAt]
+    · simp_all only [ne_eq, not_and, Decidable.not_not]
+      cases this <;> simp_all only [
+        mul_zero, one_ne_zero, zero_ne_one, sub_zero, mul_one, div_zero, sub_self]
+    · intro h
+      have contAt := h.continuousAt
+      cases this <;> simp_all [
+        not_continuousAt_deriv_qaryEntropy_zero, not_continuousAt_deriv_qaryEntropy_one, contAt]
+
+lemma deriv2_binaryEntropy {x : ℝ} : deriv^[2] binaryEntropy x = -1 / (x * (1-x)) :=
+  deriv2_qaryEntropy
+
+/-! ### Strict Monotonicity of binary entropy -/
+
+/-- Qary entropy is strictly increasing in the interval [0, 1 - q⁻¹]. -/
+lemma qaryEntropy_strictMono {q : ℕ} (qLe2: 2 ≤ q) :
+    StrictMonoOn (qaryEntropy q) (Icc 0 (1 - 1/q)) := by
+  intro p1 hp1 p2 hp2 p1le2
+  apply strictMonoOn_of_deriv_pos (convex_Icc 0 (1 - 1/(q:ℝ))) _ _ hp1 hp2 p1le2
+  · apply qaryEntropy_continuous.continuousOn
+  · intro x hx
+    have : 2 ≤ (q : ℝ) := Nat.ofNat_le_cast.mpr qLe2
+    have zero_le_qinv : 0 < (q : ℝ)⁻¹ := by positivity
+    have : 1 - x ∈ Ioi 0 := by
+      simp [mem_Ioi, sub_pos, hx.2]
+      have x_lt_1_minus_qinv : x < 1 - (q : ℝ)⁻¹ := by
+        simp_all only [inv_pos, interior_Icc, mem_Ioo, one_div]
+      linarith
+    simp only [one_div, interior_Icc, mem_Ioo] at hx
+    rw [deriv_qaryEntropy (by linarith)]
+    · field_simp
+      rw [← log_mul (by linarith) (by linarith)]
+      apply Real.strictMonoOn_log (mem_Ioi.mpr hx.1)
+      · simp_all only [mem_Ioi, mul_pos_iff_of_pos_left, show 0 < (q : ℝ) - 1 by linarith]
+      · have qpos : 0 < (q : ℝ) := by positivity
+        have : q * x < q - 1 := by
+          convert (mul_lt_mul_left qpos).2 hx.2 using 1
+          simp only [mul_sub, mul_one, isUnit_iff_ne_zero, ne_eq, ne_of_gt qpos, not_false_eq_true,
+            IsUnit.mul_inv_cancel]
+        linarith
+    exact (ne_of_gt (lt_add_neg_iff_lt.mp this : x < 1)).symm
+
+/-- Binary entropy is strictly increasing in interval [0, 1/2]. -/
+lemma binaryEntropy_strictMono : StrictMonoOn binaryEntropy (Icc 0 2⁻¹) := by
+  rw [show Icc (0:ℝ) 2⁻¹ = Icc 0 (1 - 1/2) by norm_num]
+  exact qaryEntropy_strictMono (by rfl)
```

Comment:
Can you prove it's strictly antitone on `Icc 2⁻¹ 1`?

Reply:
Done.